WIP pca_feature_visualization

cjmorale opened the pull request:
I would like to help out with the PCA Feature Visualization. I worked on some initial code today, below is a plot the code creates using the iris data set. Would love to get some feed back.

```
    from yellowbrick.features.pca import PCA2D
    from sklearn import datasets
    iris = datasets.load_iris()
    X = iris.data
    y = iris.target

    params = {'scale': True, 'center': True, 'col': y}
    visualizer = PCA2D(**params)
    visualizer.fit(X)
    visualizer.transform(X)
    visualizer.poof()
```
![iris_pca_plot](https://cloud.githubusercontent.com/assets/16965181/26291308/565bb4e2-3e62-11e7-80fa-c24f1909cf12.png)

Current features
   - User can choose to center and scale data

Needs to be completed
   - documentation
   - test cases

bbengfort commented:
@cjmorale awesome, thank you! Are you still going to be around for the sprints tomorrow (@rebeccabilbro mentioned she spoke to you at PyCon); if so we can do the code review together; otherwise I'll plan to do it tomorrow morning.

cjmorale commented:
@bbengfort I will be leaving for the airport around 9, so if you are an early riser we could do the code review together.

cjmorale commented:
Committed some new code. There is now an ipynb example if the folder `examples/cjmorale`

rebeccabilbro commented:
@cjmorale - excellent, taking a look now!

bbengfort commented on yellowbrick/features/pca.py:
```diff
@@ -0,0 +1,103 @@
+##########################################################################
+##2D PCA Visualizer
+##########################################################################
+class PCA2D(DataVisualizer):
```
This is _exactly_ what we hoped to see from new visualizers - good naming, good documentation/docstrings, tests; it's the whole package!

cjmorale replied:
Thanks!

bbengfort commented on tests/test_features/test_pca.py:
```diff
@@ -0,0 +1,71 @@
+
+class PCA2DTests(unittest.TestCase, DatasetMixin):
+
+    def test_scale_t_center_t(self, scale=True, center=True):
```
Our test package requires a one line description of the tests, e.g. 

```python
def test_scale_t_center_t(self):
    """
    Test the PCA2D visualizer scaled and centered 
    """
```

Because this line in the docstring is printed in the test report. 

Also, in terms of arguments to the test, e.g. (scale=True, center=True), our test runner won't be able to modify the arguments, so they're there for documentation only, and I think the name of the method does a good job of documenting it if you want to remove them. Otherwise you could create a function that takes the arguments that is not named `test_` and then simply pass the arguments to the function from your two test cases.

cjmorale replied:
I removed the arguments and added a one line docstrings to the functions. 

I like to use `pytest.raises` to write test cases that should catch thrown exceptions. 

 Any thoughts about adding `pytest` to the requirements.txt?

bbengfort commented on yellowbrick/features/pca.py:
```diff
@@ -0,0 +1,103 @@
+    >>> visualizer.poof()
+
+
+    Notes
```
Not necessary to include a Notes header if you don't have any!

bbengfort commented on yellowbrick/features/pca.py:
```diff
@@ -0,0 +1,103 @@
+    
+    center : bool, default: True
+    
+    col : list or tuple, default: None
```
In other visualizers we call this argument "color" -- e.g. in [ParallelCoordinates](https://github.com/cjmorale/yellowbrick/blob/feature/pca_feature_visualizer/yellowbrick/features/pcoords.py#L59); any reason not to maintain this here?

cjmorale replied:
Good point. updated line 33 to 

```{python}
    color : list or tuple of colors, default: None
        Specify the colors for each individual class
```

bbengfort commented on yellowbrick/features/pca.py:
```diff
@@ -0,0 +1,103 @@
+    
+    """
+    def __init__(self, ax=None, scale=True, center=True, col=None,
+                 colormap='RdBu_r', **kwargs):
```
We do have a default colormap, that we'd prefer you'd use, [`DEFAULT_SEQUENCE`](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/yellowbrick/style/palettes.py#L347); it's actually `RdBu` -- is reversed important? If so, keep this as it is!

cjmorale replied:
That is good to know. I am still getting familiar with the code base. Added 
`from yellowbrick.style import palettes` at imports and replaced line 64-65 with

```{python}
    def __init__(self, ax=None, scale=True, col=None, proj_dim=2,
                 colormap=palettes.DEFAULT_SEQUENCE, **kwargs):
```

bbengfort commented on yellowbrick/features/pca.py:
```diff
@@ -0,0 +1,103 @@
+        if self.ax is None:
+            self.ax = self.gca()
+
+        self.ax.scatter(X[:, 0], X[:, 1], c=self.col, cmap=plt.cm.Paired)
```
In the init you save a colormap, but don't use it here, instead using the Paired map; any reason?

cjmorale replied:
I intended to have `cmap=self.colormap`. Good eye!

bbengfort commented on yellowbrick/features/pca.py:
```diff
@@ -0,0 +1,103 @@
+    def draw(self, **kwargs):
+        X = self.pca_features_
+        if self.ax is None:
+            self.ax = self.gca()
```
In a recent commit, we changed our axes management; so `if self.ax is None: self.ax = self.gca()` is no longer necessary; you can remove these two lines of code!

cjmorale replied:
I might need some help on this one. I tried to update my branch but when I deleted those two lines the code would run completely :/

bbengfort replied:
Hmm, I went into your branch and it looks like you have the latest code. These two lines of code if they're in should cause an error at this point. Can you describe the error you were seeing?

cjmorale replied:
@bbengfort it looks like the my code runs now after excluding lines 90-91. So that issue is now resolved! Below are two errors I am getting when I run `make test`

I am trying to wrap up some of the loose ends n my code tonight. The function now supports 3d plots and I am making test cases for the *quick method* you mentioned in your code review.

```{python}
ERROR: Test radviz on the real, occupancy data set
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/carlo/Documents/yellowbrick/tests/test_features/test_radviz.py", line 89, in test_integrated_radviz
    visualizer.fit_transform(X, y)
  File "/Users/carlo/Documents/yellowbrick/myenv/lib/python3.5/site-packages/sklearn/base.py", line 497, in fit_transform
    return self.fit(X, y, **fit_params).transform(X)
  File "/Users/carlo/Documents/yellowbrick/yellowbrick/features/base.py", line 193, in fit
    self.draw(X, y, **kwargs)
  File "/Users/carlo/Documents/yellowbrick/yellowbrick/features/radviz.py", line 226, in draw
    self.ax.text(xy[0] + 0.025, xy[1] + 0.025, name, ha='left', va='bottom', size='small')
TypeError: text() missing 1 required positional argument: 's'


ERROR: Assert no errors occur during radviz visualizer integration
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/carlo/Documents/yellowbrick/tests/test_features/test_radviz.py", line 71, in test_radviz
    visualizer.fit_transform(self.X, self.y)
  File "/Users/carlo/Documents/yellowbrick/myenv/lib/python3.5/site-packages/sklearn/base.py", line 497, in fit_transform
    return self.fit(X, y, **fit_params).transform(X)
  File "/Users/carlo/Documents/yellowbrick/yellowbrick/features/base.py", line 193, in fit
    self.draw(X, y, **kwargs)
  File "/Users/carlo/Documents/yellowbrick/yellowbrick/features/radviz.py", line 226, in draw
    self.ax.text(xy[0] + 0.025, xy[1] + 0.025, name, ha='left', va='bottom', size='small')
TypeError: text() missing 1 required positional argument: 's'
```

bbengfort commented:
@cjmorale we've been updating the code base, please make sure you hit the update branch button when you're ready to make sure that the tests pass for the 0.4.2 release. For example, the axes thing I mentioned in the code review depends on it.

cjmorale commented:
@rebeccabilbro @bbengfort I pushed some new code. I tried to address all of the code comments by @bbengfort. I added more test cases and also added the quick function. The function supports 2 and 3 dimensional plots. I added some examples in the example folder see the notebook at `examples/cjmorale/PCA_Examples.ipynb`

bbengfort commented on requirements.txt:
```diff
@@ -4,6 +4,7 @@ scipy>=0.18
 scikit-learn>=0.18
 numpy>=1.11.0
 cycler>=0.10.0
+pytest>=3.0.7
```
We're really kind of wary of adding too many dependencies to Yellowbrick; we've had a tough time with the dependencies we do have in terms of getting a Conda package together, and dealing with different versions. At the very least, I'd ask you to move this dependency to `tests/requirements.txt` -- however, the way I see you using pytest below, there is actually already a solution using `unittest`; let me know what you think.

bbengfort commented on tests/test_features/test_pca.py:
```diff
@@ -0,0 +1,233 @@
+
+        params = {'scale': True, 'center': False, 'proj_dim': 4, 'col': y}
+
+        with pytest.raises(ValueError) as exception:
```
This line can be easily replaced by:

```python
with self.assertRaises(ValueError) as exception:
    ... 
```

bbengfort commented on tests/test_features/test_pca.py:
```diff
@@ -0,0 +1,233 @@
+
+        params = {'scale': True, 'center': False, 'proj_dim': 3, 'col': y}
+
+        with pytest.raises(ValueError) as exception:
```
This line can be easily replaced by:

```python
with self.assertRaises(ValueError) as exception:
    ... 
```

bbengfort commented on tests/test_features/test_pca.py:
```diff
@@ -0,0 +1,233 @@
+import unittest
+import numpy as np
+import numpy.testing as npt
+import pytest
```
Recommend removing this dependency, see below.

bbengfort commented on yellowbrick/features/pca.py:
```diff
@@ -0,0 +1,176 @@
+        super(PCADecomposition, self).__init__(ax=ax, **kwargs)
+        # Data Parameters
+        if proj_dim not in (2, 3):
+            raise ValueError("proj_dim object is not 2 or 3.")
```
It would be my preference that you raised a `YellowbrickValueError` (from the exceptions.py module) so that we can disambiguate between Yellowbrick exceptions and other exceptions. Note that `YellowbrickValueError` extends both `YellowbrickException` and `ValueError` so you can test for both as you're doing in the tests.
